azurerm/common_hdinsight.go

```diff
@@ -57,6 +60,50 @@ func hdinsightClusterUpdate(clusterKind string, readFunc schema.ReadFunc) schema
 			}
 		}
 
+		// The API can add an edge node but can't remove them without force newing the resource. We'll check for adding here
+		// and can come back to removing if that functionality gets added. https://feedback.azure.com/forums/217335-hdinsight/suggestions/5663773-start-stop-cluster-hdinsight?page=3&per_page=20
+		if clusterKind == "Hadoop" {
+			if d.HasChange("roles.0.edge_node") {
+				log.Printf("[DEBUG] Detected change in edge nodes")
+				edgeNodeRaw := d.Get("roles.0.edge_node").([]interface{})
+				edgeNodeConfig := edgeNodeRaw[0].(map[string]interface{})
+				applicationsClient := meta.(*ArmClient).HDInsight.ApplicationsClient
+
+				oldEdgeNodeCount, newEdgeNodeCount := d.GetChange("roles.0.edge_node.0.target_instance_count")
+				oldEdgeNodeInt := oldEdgeNodeCount.(int)
+				newEdgeNodeInt := newEdgeNodeCount.(int)
+
+				// Note: API currently doesn't support updating number of edge nodes
+				// if anything in the edge nodes changes, delete edge nodes then recreate them
```

Comment:
It looks like we're only destroying and recreating when `target_instance_count` changes but this comment notes that any changes will destroy and recreate an edge node. We should also check for if `install_script_actions`, the attributes in `install_script_actions`, and `vm_size` change and also trigger the destroy and recreate.

Reply:
@mbfrahry I have tested that and it does work as expected. There is no application update API, so any changes to edge nodes require a delete followed by an install.

The code there that explicitly checks on changes in the target instance count is there to avoid deleting 0 edge nodes, and creating 0 edge nodes.